ons_alpha/private_media/models.py

```diff
@@ -0,0 +1,217 @@
+                try:
+                    file_acl = file.obj.Acl()
+                except ClientError:
+                    # TODO: differentiate between general connection failures
+                    # and other issues that couldn't possibly result in success
+                    logger.exception("Failed to retreive ACL for file: %s.", repr(file))
+                    failures += 1
+                    continue
+                try:
+                    file_acl.put(ACL=acl_name)
+                except ClientError:
+                    # TODO: differentiate between general connection failures
+                    # and other issues that couldn't possibly result in success
+                    logger.exception("Failed to set ACL for file: %s.", repr(file))
+                    failures += 1
+                    continue
+
+        if failures:
+            return False
+
+        self.acls_last_set = timezone.now()
+        return True
+
+    def get_privacy_controlled_files(self) -> Iterable[File]:
+        raise NotImplementedError
+
+    @classmethod
+    def bulk_make_public(cls, objects: list["PrivateMediaMixin"]) -> int:
+        """
+        Make a list of objects of this type 'public' as efficiently as
+        possible. Returns the number of objects that were actually updated
+        in response to the request.
+        """
+        to_update = []
+        for obj in objects:
+            if obj.move_to_public_collection():
+                obj.determine_privacy()
+                to_update.append(obj)
+        if not to_update:
+            return 0
+
+        count = cls.objects.bulk_update(to_update, fields=cls.privacy_fields)
+        cls.bulk_set_file_acls(to_update, "public")
+        return count
+
+    @classmethod
+    def bulk_make_private(cls, objects: list["PrivateMediaMixin"]) -> int:
+        """
+        Make a list of objects of this type 'private' as efficiently as
+        possible. Returns the number of objects that were actually updated
+        in response to the request.
+        """
+        to_update = []
+        for obj in objects:
+            if obj.move_to_private_collection():
+                obj.determine_privacy()
+                to_update.append(obj)
+        if not to_update:
+            return 0
+
+        count = cls.objects.bulk_update(to_update, fields=cls.privacy_fields)
+        cls.bulk_set_file_acls(to_update, "private")
+        return count
+
+    @classmethod
+    def bulk_set_file_acls(cls, objects: list["PrivateMediaMixin"], acl_name: str) -> int:
+        """
+        For a list of objects of this type, set the ACLs for all related
+        files to either 'public' or 'private'. Returns the number of objects
+        for which all related files were successfully updated - which will
+        also have their 'acls_last_set' datetime updated.
+        """
+        successfully_updated = []
+        for obj in objects:
+            if obj.set_file_acls(acl_name):
```

Comment:
Question: Does `boto3` have an API to bulk update ACLs? I strongly suspect any performance issues will be from updating the ACLs in S3 rather than updating the DB.

Reply:
it has batch operations:
https://boto3.amazonaws.com/v1/documentation/api/1.26.103/reference/services/s3control/client/create_job.html
https://aws.amazon.com/blogs/storage/updating-amazon-s3-object-acls-at-scale-with-s3-batch-operations/

Reply:
@RealOrangeOne I was hoping for such a thing too, but couldn't find one unfortunately.

Reply:
Am going to leave this for now, as triggering batch operations seems way more complicated than it's worth